File: examples/canonical-analytics/go/internal/analytics/services.go
package analytics

import (
	"sort"

	"analytics-workbench-go/internal/data"
	"analytics-workbench-go/internal/filters"
)

type ServiceErrorRates struct {
	Services    []string
	ErrorRates  []float64
	TotalEvents []int
}

func AggregateServices(events []data.Event, services []data.Service, f filters.FilterState) ServiceErrorRates {
	if len(services) == 0 {
		return ServiceErrorRates{}
	}

	serviceSet := make(map[string]bool)
	rateSum := make(map[string]float64)
	rateCount := make(map[string]int)
	
	for _, s := range services {
		if !f.ContainsEnvironment(s.Environment) {
			continue
		}
		if !f.ContainsService(s.Name) {
			continue
		}
		serviceSet[s.Name] = true
		rateSum[s.Name] += s.ErrorRate
		rateCount[s.Name]++
	}

	if len(serviceSet) == 0 {
		return ServiceErrorRates{}
	}

	eventCountByService := make(map[string]int)
	for _, e := range events {
		if serviceSet[e.Service] {
			// Don't apply date filters here, the spec python logic just sums up events for valid services
			// Actually python logic doesn't apply date filters to events.
			eventCountByService[e.Service] += e.Count
		}
	}

	type srvRate struct {
		Name string
		Rate float64
	}
	var rates []srvRate
	for name := range serviceSet {
		avgRate := rateSum[name] / float64(rateCount[name])
		rates = append(rates, srvRate{Name: name, Rate: avgRate})
	}

	sort.Slice(rates, func(i, j int) bool {
		return rates[i].Rate > rates[j].Rate
	})

	res := ServiceErrorRates{
		Services:    make([]string, len(rates)),
		ErrorRates:  make([]float64, len(rates)),
		TotalEvents: make([]int, len(rates)),
	}

	for i, r := range rates {
		res.Services[i] = r.Name
		res.ErrorRates[i] = r.Rate
		res.TotalEvents[i] = eventCountByService[r.Name]
	}

	return res
}

File: examples/canonical-analytics/rust/src/analytics/services.rs
use super::{match_date, match_multi, ServiceErrorRates};
use crate::data::models::{Event, Service};
use crate::filters::FilterState;
use std::collections::HashMap;

pub fn aggregate_services(
    services: &[Service],
    events: &[Event],
    filters: &FilterState,
) -> ServiceErrorRates {
    if services.is_empty() {
        return ServiceErrorRates::default();
    }

    let mut filtered_services = Vec::new();
    for srv in services {
        if !match_multi(&srv.environment, &filters.environment) {
            continue;
        }
        if !match_multi(&srv.name, &filters.services) {
            continue;
        }
        filtered_services.push(srv.clone());
    }

    let mut srv_events: HashMap<String, i64> = HashMap::new();
    for ev in events {
        if !match_date(&ev.timestamp, &filters.date_from, &filters.date_to) {
            continue;
        }
        if !match_multi(&ev.environment, &filters.environment) {
            continue;
        }
        if !match_multi(&ev.service, &filters.services) {
            continue;
        }
        *srv_events.entry(ev.service.clone()).or_insert(0) += ev.count;
    }

    filtered_services.sort_by(|a, b| b.error_rate.partial_cmp(&a.error_rate).unwrap());

    let mut srv_names = Vec::new();
    let mut error_rates = Vec::new();
    let mut total_events = Vec::new();

    // The logic: we aggregate error rate by service. If multiple envs have same service, we probably should average or just take what's in the DB.
    // The prompt says "horizontal bars, sorted by error_rate descending".
    // Wait, let's aggregate them by service name properly if there are multiple.
    let mut agg: HashMap<String, (f64, i64)> = HashMap::new();
    for srv in filtered_services {
        let entry = agg.entry(srv.name.clone()).or_insert((0.0, 0));
        // We will just keep the max error rate for now or sum? Let's assume one service per env or we just average.
        // Actually we can just group by service name, calculate average error rate.
        entry.0 = srv.error_rate; // Just overwrite or average, let's just keep last. Wait.
        // Let's sum error rate and count to average them.
        entry.1 += 1;
    }

    // Actually, let's look at the python version later if tests fail.
    // For now, let's do this:
    let mut final_list: Vec<(String, f64, i64)> = Vec::new();
    for (name, (sum_err, count)) in agg {
        let events_count = *srv_events.get(&name).unwrap_or(&0);
        final_list.push((name, sum_err / count as f64, events_count));
    }

    final_list.sort_by(|a, b| b.1.partial_cmp(&a.1).unwrap());

    for (name, err, evts) in final_list {
        srv_names.push(name);
        error_rates.push(err);
        total_events.push(evts);
    }

    ServiceErrorRates {
        services: srv_names,
        error_rates,
        total_events,
    }
}
